### PR TITLE
Header updates - breaking changes

### PR DESCRIPTION
**Pull request by anandamaryon1**
## Description
Bringing together a series of breaking changes from the following PRs:
- [ ] https://github.com/nhsuk/nhsuk-frontend/pull/1047
- [ ] https://github.com/nhsuk/nhsuk-frontend/pull/1000
- [ ] https://github.com/nhsuk/nhsuk-frontend/pull/996

## Checklist

- [ ] Tested against our [testing policy](https://github.com/nhsuk/nhsuk-frontend/blob/master/docs/contributing/testing.md) (Resolution, Browser & Accessibility)
- [ ] Follows our [coding standards and style guide](https://github.com/nhsuk/nhsuk-frontend/blob/master/docs/contributing/coding-standards.md)
- [ ] CHANGELOG entry

### CODE REVIEW

**Comment by anandamaryon1** [comment]
@paulrobertlloyd found an issue with your logo SVG change (https://github.com/nhsuk/nhsuk-frontend/pull/1047), when printing the NHS logo no longer shows. Not looked into the code but I assume it's just missing a rule to change it's colour from white to blue on print?

**Comment by frankieroberto** [comment]
@paulrobertlloyd could we avoid force pushing to this branch for a bit? Makes it harder to maintain the other pull requests into this branch. We can sort out the commit history at the end?

**Comment by anandamaryon1** [comment]
Potential new issue to include in the PR: Making the header work without JS enabled and progressively enhancing the responsive layout via JS. 

Currently the responsive layout without JS is a bit broken, with the nav links overflowing rather than wrapping.

**Comment by anandamaryon1** [comment]
# Accessibility and device testing

Tested with NVDA, JAWS and VoiceOver, all work as expected.

Tested on Safari on Mac, and Edge and Chrome Windows. Android devices Galaxy 21 Samsung browser and Pixel 6 Chrome, all as expected. Tested on iOS Safari, all as expected on version 15 and above. 

## Issues
Two issues discovered on Safari 14 (iPhone 8, iPad mini 2019 & iPad air 4) and below:

1. Flexbox `gap` is not supported.
2. Search input gets styling applied to round it's corners. 

These issues do not affect the current header implementation.

image
image
image

Raises a wider question of our browser support which has not been updated in 3 years. https://github.com/nhsuk/nhsuk-service-manual/issues/2063

**Comment by davidhunter08** [comment]
Following on from a previous [comment on underlining the nav links](https://github.com/nhsuk/nhsuk-frontend/pull/1067#issuecomment-2465281775), I noticed [gov guidance on underlining links](https://design-system.service.gov.uk/styles/links/#links-without-underlines)...

> links in a header or side navigation might not need underlines. Users will understand that they’re links because of where they are on the page.

I mocked up a header for the NHS App (using the new 'account info' variant)...

header-as-is

I thought I'd see how it looked without underlines (and removed the `justify-content: space-between;`) and it feels cleaner/more legible IMO...

header-proposal

**Comment by anandamaryon1** [comment]
@davidhunter08 
Comparing side by side, it does look a lot cleaner with less visual noise. 

Horizontal justification of the links also helps in your example. It works well on nhs.uk due to the number and length of links, but as a rule for all headers it may not make so much sense. 

If we're going to do it, this is the time and this is the PR.

Thoughts @frankieroberto @paulrobertlloyd ?

**Comment by paulrobertlloyd** [comment]
- Maybe we only underline navigation links on hover? It’d be good to have some way of differentiating between things that are links and those that are not (this largely applies to the account area, where some items may be text only)
- I’d love for the justification of the links to no longer be the default; maybe we can provide a class to enable it?

**Comment by frankieroberto** [comment]
➕1 for removing the underlines from the horizontal nav (except maybe on hover) and removing the default justification.

It would also fix the currently-misleading issue that the 'More' toggle button uses an underline, and so might be mistaken for a link.

**Comment by paulrobertlloyd** [comment]
### Issues found during accessibility and device testing

@anandamaryon1 I looked at how we might use margins instead for navigation items instead of flexbox in #1138, but it got complicated quickly. I suspect the same is true for other areas where the design falls apart a little due to its reliance on using `gap` with flexbox.

I think the bunched up navigation items are the primary concern, one which was mitigated against previously because items had generous padding to their left and right. With that padding removed, they will line up closer to each other.

Safari on iOS is a pain because users can’t upgrade to a newer browser, and all browsers, updated or not, are forced to use the same underlying Webkit rendering engine. Once OS updates are no longer provided, the user is stuck with the last supported version of WebKit, which in the above case is 14.

This is even more so annoying because this version supports grid layout and therefore `gap/column-gap/row-gap` properties, but does _not_ support using these properties for Flexbox containers. I’m not sure if we can test for this, even by using `@supports` feature queries.

Curious to hear if others have suggestions for how/if we can fix the above noted issues for older browsers. @frankieroberto @colinrotherham

**Comment by anandamaryon1** [comment]
## DSRP account header feedback summary
The design system review panel has reviewed this component and provided the following feedback, which we will look to address before releasing.

### Component design (frontend)
- Is the dark blue background on account links necessary? Does it make the account links overly prominent?
- Ensure very long emails do not break the layout
- Review wrapping of elements within custom HTML items

### Component guidance (service manual)
- Should there be more guidance on how to display user names, roles etc.?
- Should the log out button be optional or responsively shown, worry it could be error-prone on mobile
- If using, the new dark blue background colour needs to be added to design system, or an existing colour should be used instead
- Emphasise the fact that this has been designed for staff-facing systems, consider releasing as experimental
- Add further guidance to encourage testing to inform account header information and links to be displayed, eg. "user testing should inform which combination of elements will best meet user needs"
- Highlight the fact that items can be links and non-links (information)
- There could be potential for a future pattern for changing role/organisation
- Provide guidance on search placeholder text

**Comment by paulrobertlloyd** [comment]
@anandamaryon1 Have updated and merged #1138 and #1155 into this branch as discussed earlier this week. @frankieroberto; shout if you need a hand rebasing these changes into #1109.

**Comment by frankieroberto** [comment]
@paulrobertlloyd @anandamaryon1 @colinrotherham I’ve merged in #1109 which adds the breaking changes to the Nunjucks params.

The visual regression tests are now failing on this PR, but they’re also failing for me when running locally on `main` so I’m not quite sure what the issue is.

I’ve added a Todo to the PR description to consider whether we want to make further breaking changes to the CSS names for the header to bring them more in line with BEM conventions, eg changing `.nhsuk-navigation` to `.nhsuk-header__navigation`. I could be persuaded either way on this, but I guess if we are going to change them, it probably makes sense to do it as part of this breaking change release?

We’ll also need to write a ton of guidance on how to switch over to the new Nunjucks param names, either in the CHANGELOG or the release notes (or both).

**Comment by paulrobertlloyd** [comment]
Some of the failing tests seem to be environment specific (namely the typeface rendering differently), which I wonder if it’s a case of regenerating the screenshots.

But there does appear to be a few genuine regressions. I can take a look at those now.

**Comment by paulrobertlloyd** [comment]
@frankieroberto Fixed the following:

- Organisation logo no-longer incorrectly wraps
- Short service name now stays to the right of the NHS logo on narrower viewports
- NHS logo appears black when printed
- Made the template (hopefully) easier to read by moving the logic for the logo and service name into their own macros.

Also updated the HTML. There’s now always a `nhsuk-header__service` container which dictates the behaviour of this part of the header (as opposed to having each individual piece do this, contextually). And I’ve slightly renamed things as follows:

- `div.nhsuk-header__service`
  - `a.nhsuk-header__service-logo`
    - [`span.nhsuk-header__service-name`]
  - [`a|span.nhsuk-header__service-name`]

(In two minds whether to rename `.nhsuk-header__service-logo` to something like `.nhsuk-header__service-link` or `.nhsuk-header__service-home`… not sure; this was previously `.nhsuk-header__logo-link`. Maybe that name was fine, but wanted to use `service` in the name so can show the relationship between these elements.)

**Comment by frankieroberto** [comment]
@paulrobertlloyd nice! I struggled with the css names too, especially for the link which can sometimes contain the logo or service name, and sometimes both together… 😬

Shall we go the whole hog and rename `.nhsuk-navigation` too? 🤔

May also be time to move all the header css into a single file, although it would be a long one.

**Comment by paulrobertlloyd** [comment]
> Shall we go the whole hog and rename .nhsuk-navigation too?

I’d be tempted to actually have a separate navigation component (aka ‘service navigation’?) and have it use a separate namespace, CSS, JS and Nunjucks etc., even though this could be presented as one component as it is now. That way these things would be easier to manage and reason with. (There is precedent for this; things like hint text and labels)

**Comment by paulrobertlloyd** [comment]
Added another to-do item; the unit tests for the header navigation appear to be run against markup that is no longer used, so should be updated. For example, it assumes certain parts of the header markup are added via JS, but in fact now provided by default and then hidden by CSS.

I could be wrong, but as I was playing with #1193, this is how it looked to me.

**Comment by frankieroberto** [comment]
@paulrobertlloyd:
> I’d be tempted to actually have a separate navigation component (aka ‘service navigation’?) and have it use a separate namespace, CSS, JS and Nunjucks etc., even though this could be presented as one component as it is now.

I'm not so sure about this - the navigation feels like an integral part of the header to me, unlike the govuk one, where you might even have some components between the header and the service navigation (eg language switchers or something).

Hint text and labels are a precedent for nested components, but maybe they’re justified on the grounds of being used in more than 1 component, unlike the navigation? 🤔

**Comment by anandamaryon1** [comment]
I think let's hold off on splitting out navigation for now, to contain this ever-expanding PR :'D 

Sorting out the CSS to be BEM makes sense I suppose, within this breaking change, and I think we're doing similar with the table right now, right @colinrotherham ?

**Comment by anandamaryon1** [comment]
Added `overflow-wrap: anywhere;` to the `.nhsuk-header__account-button, .nhsuk-header__account-link` classes, to prevent very long names and email addresses from breaking the layout on smaller screens.

File: packages/components/header/_header.scss [hunk 1]
```diff
@@ -125,6 +125,7 @@
   @include nhsuk-link-style-header;
   display: flex;
   gap: nhsuk-spacing(2);
+  overflow-wrap: anywhere;
```
[the remark this answers is not in the document]
Comment (anandamaryon1) [reply packages/components/header/_header.scss:128]: I did try that first, but alas it didn't work. Something to do with flexbox and nested containers.

File: CHANGELOG.md [hunk 2]
```diff
@@ -4,6 +4,64 @@
...
+    </li>
+    ...
+  </ul>
+  ```
+
+  If you are using the `.nhsuk-header__navigation-list--left-aligned` modifier class, this can now be removed.
+
+:boom: **Breaking changes**
+
+- Remove the boolean `showNav`, `showSearch` and `transactional` options from the header component. Respective parts of the header are now shown if values for `primaryLinks`, `search` or `transactionalService` options are provided. Additionally, the `searchAction` option is renamed `search.action` and the `searchInputName` option is renamed `search.name`. Finally, the label, button and placeholder text for the search input can be updated using the new `search.visuallyHiddenLabel`, `search.visuallyHiddenButton` and `search.placeholder` options. ([PR 996](https://github.com/nhsuk/nhsuk-frontend/pull/996))
```
[the remark this answers is not in the document]
Comment (anandamaryon1) [reply CHANGELOG.md]: Nice and clear 🙏

File: CHANGELOG.md [hunk 3]
```diff
@@ -4,6 +4,64 @@
...
+
+  This has been updated to use the following SVG:
+
+  ```svg
+  <svg class="nhsuk-logo" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 200 80" height="40" width="100">
+    <path fill="currentcolor" d="M200 0v80H0V0h200Zm-27.5 5.5c-14.5 0-29 5-29 22 0 10.2 7.7 13.5 14.7 16.3l.7.3c5.4 2 10.1 3.9 10.1 8.4 0 6.5-8.5 7.5-14 7.5s-12.5-1.5-16-3.5L135 70c5.5 2 13.5 3.5 20 3.5 15.5 0 32-4.5 32-22.5 0-19.5-25.5-16.5-25.5-25.5 0-5.5 5.5-6.5 12.5-6.5a35 35 0 0 1 14.5 3l4-13.5c-4.5-2-12-3-20-3Zm-131 2h-22l-14 65H22l9-45h.5l13.5 45h21.5l14-65H64l-9 45h-.5l-13-45Zm63 0h-18l-13 65h17l6-28H117l-5.5 28H129l13.5-65H125L119.5 32h-20l5-24.5Z"/>
+  </svg>
+  ```
+
+- Remove hardcoded home link from navigation ([PR 986](https://github.com/nhsuk/nhsuk-frontend/pull/986))
```
[the remark this answers is not in the document]
Comment (anandamaryon1) [reply CHANGELOG.md:64]: Was merged into this PR. Makes a bit of a messy history though, can't recall why we didn't change the base and merge… 😞

File: packages/components/header/_header-organisation.scss [hunk 4]
```diff
@@ -114,10 +57,10 @@
...
 
-.nhsuk-organisation-logo[src$=".svg"] {
-  height: auto;
-  max-width: 220px;
-  width: 100%;
+  &[src$=".svg"] {
+    height: auto;
+    max-width: 220px;
+    width: 100%;
+  }
```
[the remark this answers is not in the document]
Comment (anandamaryon1) [reply packages/components/header/_header-organisation.scss:65]: Hmm will have to give it a test.

[the remark this answers is not in the document]
Comment (anandamaryon1) [reply packages/components/header/_header-organisation.scss:65]: Assuming I'm doing this right (which I may not be…) I can't get a PNG override image to work.

Trying this:
image

image

Results in still showing the SVG, not the linked PNG.

File: packages/components/header/_header.scss (hunk 1 above)
[the remark this answers is not in the document]
Comment (paulrobertlloyd) [reply packages/components/header/_header.scss:128]: Do you mean using `overflow-wrap: anywhere` in the mixin @colinrotherham (and _then_ using the mixin here)?

File: packages/components/header/_header-organisation.scss (hunk 4 above)
[the remark this answers is not in the document]
Comment (paulrobertlloyd) [reply packages/components/header/_header-organisation.scss:65]: This is in the existing styles: https://github.com/nhsuk/nhsuk-frontend/blob/7fc989019bd8423717c157c3374886165ba90aee/packages/components/header/_header-organisation.scss#L119C1-L123C2

Trying to think why you would need a separate declaration for image that happens to be an SVG? Definitely worth testing; if it’s still needed we should add an explanatory comment.

[the remark this answers is not in the document]
Comment (anandamaryon1) [reply packages/components/header/_header-organisation.scss:65]: Ok, tried like this: 

```
    logo: {
      href: "#",
      src: "https://upload.wikimedia.org/wikipedia/commons/thumb/d/d3/National_Health_Service_%28England%29_logo.svg/640px-National_Health_Service_%28England%29_logo.svg.png"
    },
```

And I get this error:

```
Error in plugin "validateHTML"
Message:
    error: Expected omitted end tag <img> instead of self-closing element <img/> (void-style) at dist/app/components/header/header-org.html:40:324:
```
Seems it doesn't like the self-closing `` tag.

So I took the `/` off the end of the image tag and that seems to work. (but doesn't seem right to me?)

So now I can see the PNG in there correctly: 
image

But given this is for custom organisational logos, I don't think we can provide a standard height and width, like we do for the svg. (assuming the intention is for the whole org logo, including the NHS logo at the top, is to be added as one image)

So long story short, I don't think these styles should apply to PNGs, they're correct to be SVG only. 

But an aside, should I commit the change to the `` tag, I always thought they should be ``

[the remark this answers is not in the document]
Comment (paulrobertlloyd) [reply packages/components/header/_header-organisation.scss:65]: All the gory details https://jakearchibald.com/2023/against-self-closing-tags-in-html/

[the remark this answers is not in the document]
Comment (anandamaryon1) [reply packages/components/header/_header-organisation.scss:65]: Ha, I'll need a snorkel for that one @paulrobertlloyd… 

I suppose we'll need to remove the `/` self closing tag since our tests need it then? 

I did a search and we have some `/>` in this file too: `/packages/components/error-summary/README.md` shall I fix them here now too? (to avoid a new PR?) Or is that bad practice?

[the remark this answers is not in the document]
Comment (frankieroberto) [reply packages/components/header/_header-organisation.scss:65]: Happy to remove the `/` from the `` tag - although it’s like that right now, so not sure why the `validateHTML` test isn’t picking it up? 🤷‍♂️

[the remark this answers is not in the document]
Comment (anandamaryon1) [reply packages/components/header/_header-organisation.scss:65]: Ok, removed the `/` from ``. Not searched and destroyed any other closing tags, for now.

[the remark this answers is not in the document]
Comment (paulrobertlloyd) [reply packages/components/header/_header-organisation.scss:65]: Playing with the above, seems like it’s a superfluous, and possibly unhelpful rule. Using the NHS England logo as a (bad) example:

PNG:

Screenshot 2025-04-03 at 16 39 33

SVG image (with above rule):

Screenshot 2025-04-03 at 16 38 33

SVG image (with above rule removed):

Screenshot 2025-04-03 at 16 38 56


Worth a separate PR (perhaps with examples to visually regress) to remove these styles?

**Comment by paulrobertlloyd** [comment]
> @paulrobertlloyd Mind if we check the up/down chevron vertical alignment on mobile?

Fixed in c4fecb2 (where I also move the chevron specific styles into `_header.scss`

**Comment by paulrobertlloyd** [comment]
All these links changes look good. They’ll look even better once we swap the hover and active link colours around 😉

**Review by paulrobertlloyd — APPROVED** [review]
After lots and of feedback and polish over several months, this component is in a much, much healthier state. SHIP. IT. 🚢
